wgpu-core/src/pool.rs

```diff
@@ -60,12 +68,47 @@ impl<K: Clone + Eq + Hash, V> ResourcePool<K, V> {
             //
             // We pass the strong reference outside of the closure to keep it alive while we're the only one keeping a reference to it.
             let mut strong = None;
-            let weak = entry.get_or_try_init(|| {
-                let strong_inner = constructor.take().unwrap()(key.take().unwrap())?;
+            #[cfg(not(feature = "once_cell"))]
+            let mut removal_key = None;
+            let try_constructor = || {
+                let key = key.take().unwrap();
+
+                #[cfg(not(feature = "once_cell"))]
+                {
+                    removal_key = Some(key.clone());
+                }
+
+                let strong_inner = constructor.take().unwrap()(key)?;
                 let weak = Arc::downgrade(&strong_inner);
                 strong = Some(strong_inner);
                 Ok(weak)
-            })?;
+            };
+            cfg_if::cfg_if! {
```

Comment:
This pains me a lot to have yet another layer of stuff to try to get through. I'm worried about the consequences of various program orderings in the non-once-cell branch here, so I'll need to come back and re-audit this.